File: solutions/prospective/algorithms/CLRS/js/heapsort.js
/**
 * Heapsort
 * 
 * Time Complexity: O(n log n)
 * Space Complexity: O(1)
 * 
 * Sorts an array by building a max-heap and then repeatedly extracting the max.
 */
// === MEMO START ===
function heapify(arr, n, i) {
    let largest = i;
    const l = 2 * i + 1;
    const r = 2 * i + 2;

    if (l < n && arr[l] > arr[largest]) {
        largest = l;
    }

    if (r < n && arr[r] > arr[largest]) {
        largest = r;
    }

    if (largest !== i) {
        [arr[i], arr[largest]] = [arr[largest], arr[i]];
        heapify(arr, n, largest);
    }
}

function heapSort(arr) {
    const n = arr.length;

    // Build max heap
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // Extract elements one by one
    for (let i = n - 1; i > 0; i--) {
        [arr[0], arr[i]] = [arr[i], arr[0]]; // swap
        heapify(arr, i, 0);
    }

    return arr;
}

File: solutions/prospective/algorithms/CLRS/lua/heapsort.lua
--[[
Heapsort

Time Complexity: O(n log n)
Space Complexity: O(1)

Sorts an array by building a max-heap and then repeatedly extracting the max.
--]]
-- === MEMO START ===
function heapify(arr, n, i)
    local largest = i
    local l = 2 * i
    local r = 2 * i + 1
    
    if l <= n and arr[l] > arr[largest] then
        largest = l
    end
    if r <= n and arr[r] > arr[largest] then
        largest = r
    end
    
    if largest ~= i then
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)
    end
end

function heap_sort(arr)
    local n = #arr
    
    -- Build max heap
    for i = math.floor(n / 2), 1, -1 do
        heapify(arr, n, i)
    end
    
    -- Extract elements one by one
    for i = n, 2, -1 do
        arr[1], arr[i] = arr[i], arr[1] -- swap
        heapify(arr, i - 1, 1)
    end
    
    return arr
end

